src/service/viewport/viewport-binding-ios-embed-sd.js

```diff
@@ -0,0 +1,429 @@
+    this.wrapper_ = doc.createElement('div');
+    this.scroller_.appendChild(this.wrapper_);
+    this.wrapper_.id = 'i-amphtml-body-wrapper';
+    setImportantStyles(this.wrapper_, {
+      'overflow': 'visible',
+      'position': 'relative',
+    });
+    // Other properties will be copied from the `<body>`.
+
+    /** @private @const {!Observable} */
+    this.scrollObservable_ = new Observable();
+
+    /** @private @const {!Observable} */
+    this.resizeObservable_ = new Observable();
+
+    /** @const {function()} */
+    this.boundScrollEventListener_ = this.onScrolled_.bind(this);
+
+    /** @const {function()} */
+    this.boundResizeEventListener_ = this.onResized_.bind(this);
+
+    /** @private @const {boolean} */
+    this.useLayers_ = isExperimentOn(this.win, 'layers');
+
+    /** @private {number} */
+    this.paddingTop_ = 0;
+
+    // Setup UI.
+    /** @private {boolean} */
+    this.setupDone_ = false;
+    waitForBody(doc, this.setup_.bind(this));
+
+    // Set overscroll (`-webkit-overflow-scrolling: touch`) later to avoid
+    // iOS rendering bugs. See #8798 for details.
+    whenDocumentReady(doc).then(() => {
+      // There's no way to test `-webkit-overflow-scrolling` style on most of
+      // browsers. Thus we'll use the `i-amphtml-ios-overscroll` class for
+      // testing.
+      this.scroller_.classList.add('i-amphtml-ios-overscroll');
+      setImportantStyles(this.scroller_, {
+        '-webkit-overflow-scrolling': 'touch',
+      });
+    });
+
+    dev().fine(TAG_, 'initialized ios-embed-wrapper-sd viewport');
+  }
+
+  /** @override */
+  ensureReadyForElements() {
+    this.setup_();
+  }
+
+  /** @private */
+  setup_() {
+    if (this.setupDone_) {
+      return;
+    }
+    this.setupDone_ = true;
+
+    // Embedded scrolling on iOS is rather complicated. IFrames cannot be sized
+    // and be scrollable. Sizing iframe by scrolling height has a big negative
+    // that "fixed" position is essentially impossible. The only option we
+    // found is to reset scrolling on the AMP doc, which wraps the natural BODY
+    // children inside the `overflow:auto` element inside the shadow root of the
+    // body. For reference, here are related
+    // iOS issues (Chrome issues are also listed for reference):
+    // - https://code.google.com/p/chromium/issues/detail?id=2891
+    // - https://code.google.com/p/chromium/issues/detail?id=157855
+    // - https://bugs.webkit.org/show_bug.cgi?id=106133
+    // - https://bugs.webkit.org/show_bug.cgi?id=149264
+    const doc = this.win.document;
+    const body = dev().assertElement(doc.body, 'body is not available');
+    const shadowRoot = body.attachShadow({mode: 'open'});
+
+    // Main slot will absorb all undistributed children.
+    const mainSlot = doc.createElement('slot');
+    this.wrapper_.appendChild(mainSlot);
+    shadowRoot.appendChild(this.scroller_);
+
+    // Update body styles and monitor for further changes.
+    this.updateBodyStyles_();
```

Comment:
Can you explain why this necessary? It was glossed over in the Medium article.

Reply:
Consider the structure that we currently have:

```
<body>
  #shadow
     <div scroller>
        <div body-wrapper>
           <slot></slot>
        </div>
     </div>
</body>
```

If a custom css declared `body {display: flex; flex-direction: column; align-items: center}`, this styles need to be now re-applied to body-wrapper. Otherwise, the slotted children will no longer behave as desired - as if they were laid out using flex. Hence the copy operation. Not super elegant - but it's practical.